ldap/servers/plugins/chainingdb/cb_conn_stateless.c

```diff
@@ -846,40 +846,84 @@ cb_stale_all_connections(cb_backend_instance *cb)
 int
 cb_ping_farm(cb_backend_instance *cb, cb_outgoing_conn *cnx, time_t end_time)
 {
-
+    int version = LDAP_VERSION3;
     char *attrs[] = {"1.1", NULL};
     int rc;
+    int ret;
     struct timeval timeout;
     LDAP *ld;
     LDAPMessage *result;
     time_t now;
     int secure;
-    if (cb->max_idle_time <= 0) /* Heart-beat disabled */
+    char *plain = NULL;
+    const char *target = NULL;
+
+    if (cb->max_idle_time <= 0) {
+        /* Heart-beat disabled */
         return LDAP_SUCCESS;
+    }
+
+    const Slapi_DN *target_sdn = slapi_be_getsuffix(cb->inst_be, 0);
+
+    if (target_sdn == NULL) {
+        return LDAP_NO_SUCH_ATTRIBUTE;
+    }
+
+    target = slapi_sdn_get_dn(target_sdn);
 
-    if (cnx && (cnx->status != CB_CONNSTATUS_OK)) /* Known problem */
+    if (cnx && (cnx->status != CB_CONNSTATUS_OK)) {
+        /* Known problem */
         return LDAP_SERVER_DOWN;
+    }
 
     now = slapi_current_rel_time_t();
-    if (end_time && ((now <= end_time) || (end_time < 0)))
+    if (end_time && ((now <= end_time) || (end_time < 0))) {
         return LDAP_SUCCESS;
+    }
+
+    ret = pw_rever_decode(cb->pool->password, &plain, CB_CONFIG_USERPASSWORD);
+    if (ret == -1) {
+        return LDAP_INVALID_CREDENTIALS;
+    }
 
     secure = cb->pool->secure;
     if (cb->pool->starttls) {
         secure = 2;
     }
     ld = slapi_ldap_init(cb->pool->hostname, cb->pool->port, secure, 0);
     if (NULL == ld) {
+        if (ret == 0) {
```

Comment:
Everything looks good, but you probably don't technically need this "ret" check, same with the one check down below.  But it's also technically not any less performat, since slapi_ch_free_string does a NULL check :-)  All that being said, ack.